Add optional collection home page and descriptive collection content fields

alexklbuckley opened the pull request:
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #3255

## Description
Add new collection home page to display new descriptive collection content inputs.

## Instructions for Reviewers

List of changes in this PR:
* Adds new collection-home ts, html and scss file
* Adds /home to collection-page-routes.ts
* Adds new getCollectionHomeRoute() function to collection-page-routing-paths.ts
* Links to new collection home page from collection list.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [ ] My PR **passes [ESLint](https://eslint.org/)** validation using `yarn lint`
- [ ] My PR **doesn't introduce circular dependencies** (verified via `yarn check-circ-deps`)
- [ ] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [ ] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented:
@alexklbuckley : This PR needs more information about how someone can test this feature.  It doesn't appear to have any instructions on how this "collection home page" can be used, and it's unclear (to me) how to begin testing it or what "editable content" is supposed to be displayed.

Could you please update the description of this PR to provide instructions for reviewers on how this is supposed to work?

alexklbuckley commented:
> @alexklbuckley : This PR needs more information about how someone can test this feature. It doesn't appear to have any instructions on how this "collection home page" can be used, and it's unclear (to me) how to begin testing it or what "editable content" is supposed to be displayed.
> 
> Could you please update the description of this PR to provide instructions for reviewers on how this is supposed to work?
> 
> Until those details can be added, I've placed this in the "On Hold" column of our 9.0 board.

Thanks @tdonohue , I've updated the description of this PR to provide instructions for reviewers on how to test this.

I've also noted in the description that the collection home page is optional, and is routed to only if a new collection.routeThrough.collectionHomePage  config in the config.*.yml file is set to 'true'.

Also, the new collection content fields do enable users to save HTML content. However, that data can only be saved by administrators and is handled in the same way as the collections dc.description does.

One thing I am not sure about, is how can I add a new metadata schema and then fields to that schema. Currently, in my test plan (in the PR description) steps 13 and 14 I am asking the tester to add that through DSpace Admin Registries > Metadata. But is there a better way?

alexklbuckley commented:
> @alexklbuckley : Metadata fields should be added on the backend via a metadata schema registry file in this folder: https://github.com/DSpace/DSpace/tree/main/dspace/config/registries
> 
> For instance, the entire "dc" list of fields is in https://github.com/DSpace/DSpace/blob/main/dspace/config/registries/dublin-core-types.xml
> 
> In this case, it appears you are defining a new "collection" metadata schema, which means it'd need its own "collection-types.xml" file (or similar) to define that schema. Then, it'd need to be added to the list of "registry.metadata.load" setting in dspace.cfg: https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L925-L945 (As that's the list of metadata configs which DSpace will load during installation/upgrade)
> 
> For now, your manual process seems sufficient to have others test out this PR and provide feedback. However, assuming there's agreement in this direction, we'd likely need to have a corresponding backend PR created which adds a new "collection-types.xml" file or similar. You can wait to create that secondary PR though if you wish to simply get early feedback on this PR first.

thank you for that information @tdonohue !

Yes, I will wait to create that secondary PR to get early feedback on this PR first.

I've got all the unit tests passing on this PR now.

I am not sure about the remaining checkboxes in the checklist, so any guidance on that would be very much appreciated!

mwoodiupui commented:
Since the new metadata fields are specific to DSpace, should they not go into the 'dspace' schema?  Or is this meant to be a general-purpose namespace, published for other repository products to adopt?

mwoodiupui commented:
Aside:  the registry editor in DSpace is handy for building and tweaking a new namespace, and I've been meaning to write a registry "unloader" to complement the registry loader which is run during installation (which loading is why a schema definition file is required).

tdonohue commented:
@mwoodiupui : Good point, these metadata fields likely **should** go in the `dspace` schema, as that schema is meant to represent "internal to DSpace" metadata.   You are correct that, these new collection fields seem internal to DSpace.  So, we likely should move them under `dspace.collection.*` fields, @alexklbuckley .

alexklbuckley commented:
Thanks @mwoodiupui and @tdonohue !

I've pushed a follow-up to this PR to use dspace.collection.* fields.

I've also now pushed a new PR to DSpace backend to include the 4 new metadata files into the existing 'dspace' schema registry file: https://github.com/DSpace/DSpace/pull/9794

alexklbuckley commented:
Hi @tdonohue are we able to take off the 'needs discussion' label  for this pull request?

As I have done a follow-up pull request ( https://github.com/DSpace/DSpace/pull/9794 ) addressing the feedback in https://github.com/DSpace/dspace-angular/pull/3256#issuecomment-2321472829

kshepherd commented:
@alexklbuckley just a tip about sanitisation as that will definitely come up in a review:
We have an existing method for sanitising HTML text (scripts and other potentially unsafe things removed) before display, rather than just including it straight into innerHTML, in the `markdown.directive.ts` directive -- it's possible that you could use this directive directly yourself (HTML is valid markdown already, right?) or at least reuse the block that sanitizes HTML with `DomSanitizer`.

alexklbuckley commented:
> @alexklbuckley just a tip about sanitization as that will definitely come up in a review: We have an existing method for sanitizing HTML text (scripts and other potentially unsafe things removed) before display, rather than just including it straight into innerHTML, in the `markdown.directive.ts` directive -- it's possible that you could use this directive directly yourself (HTML is valid markdown already, right?) or at least reuse the block that sanitizes HTML with `DomSanitizer`.

Hi @kshepherd thank you for your reply.

As I am new to this, can I please check, is it this method https://github.com/DSpace/dspace-angular/blob/fca5700012ec63f17b1fa645f8ae39af4f9ca122/src/app/shared/utils/markdown.directive.ts#L56C9-L56C15 that you would recommend we could use?

If yes, could you please point us to an example of how it is used directly?

Thanks so much
Alex

kshepherd commented:
The line I was referring to is at https://github.com/DSpace/dspace-angular/blob/fca5700012ec63f17b1fa645f8ae39af4f9ca122/src/app/shared/utils/markdown.directive.ts#L70

The first think I would try is just importing and re-using the markdown directive as it is, directly - even though you're passing HTML in, my understanding is this should be considered valid markdown (I could be wrong), e.g.
`<div [dsMarkdown]="customHtml"></div>` instead of directly setting `[innerHTML]` declaratively.

If that doesn't work, in your backing component, you could inject the DOMSanitizer in your constructor, and implement something similar to the dsMarkdown directive, or just a function that calls the sanitize method, like `<div [innerHTML]="sanitizeInput(customHtml))"></div>`

Where the function is like
```
public sanitizeInput(value) { 
    return this.sanitizer.sanitize(SecurityContext.HTML, value);
}
```

(this has the benefit of not trying to render markdown to HTML first).

None of this is tested, just concepts/rough templates of things to try.

kshepherd commented:
> The first thing I would try is just importing and re-using the markdown directive as it is, directly - even though you're passing HTML in, my understanding is this should be considered valid markdown (I could be wrong), e.g. `<div [dsMarkdown]="customHtml"></div>` instead of directly setting `[innerHTML]` declaratively.

Quick update to this - I just worked on a feature where I was including some copyright/attribution HTML in an element, and the `[dsMarkdown]` directive worked, although does not sanitize HTML if the markdown feature is disabled.

I just opened a quick PR to fix this, so you might find it useful for your local code, too, both as a fix for the markdown directive if you plan to use it, or a direct example of using DOM sanitizer: https://github.com/DSpace/dspace-angular/pull/3375

alexklbuckley commented:
Unit tests are fixed

alexklbuckley commented:
Hi @tdonohue and @kshepherd as we have sanitization of html in place now and all the unit tests are passing are we able to remove the 'needs discussion' label from this pull request?

alexklbuckley commented:
Hi @tdonohue and @kshepherd  , checking in what the next steps for this enhancement are?

What would you suggest we do to get this upstreamed?

tdonohue commented:
@alexklbuckley : The next step is for this PR and #3250 is to discuss this approach among more of our Developers team (and especially Committers).  This is the first step of PRs that starts to bring DSpace into CMS-like activities for Collections (only), and we need to have a broader discussion about that direction & the approach of storing this data as metadata.

To be clear, that doesn't mean this will be rejected.  It just means that you've done something a bit unexpected, but still _interesting_.  We need to talk about it in more detail before it's merged, especially to ensure this is useful to others and that no Committers see issues with adding basic CMS functions to these pages.

This has been added as [an "Upcoming Topic" for a future Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/2025-01-09+DSpace+Developers+Meeting#id-20250109DSpaceDevelopersMeeting-Agenda)...but I haven't set a date for it yet, as many are still returning from holidays.

alexklbuckley commented:
> @alexklbuckley : The next step is for this PR and #3250 is to discuss this approach among more of our Developers team (and especially Committers). These are the first PRs that start to bring DSpace into CMS-like activities for Collections (only), and we need to have a broader discussion about that direction & the approach of storing this data as metadata.
> 
> To be clear, that doesn't mean this will be rejected. It just means that you've done something a bit unexpected, but still _interesting_. We need to talk about it in more detail before it's merged, especially to ensure this is useful to others and that no Committers see issues with adding basic CMS functions to these pages.
> 
> This has been added as [an "Upcoming Topic" for a future Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/2025-01-09+DSpace+Developers+Meeting#id-20250109DSpaceDevelopersMeeting-Agenda)...but I haven't set a date for it yet, as many are still returning from holidays.

Thanks very much @tdonohue , appreciate your reply and adding this to the upcoming topic list for a future developer meeting.

We look forward to hearing what the community think

tdonohue commented:
I realize I forgot to update this PR with the discussion that took place on [Jan 16 in our Developers Meeting](https://wiki.lyrasis.org/display/DSPACE/2025-01-16+DSpace+Developers+Meeting).  

There was some general interest in bringing features to the User Interface that allow Administrators more control over the User Interface look & feel.  However, there were concerns about allowing *editing of CSS directly* (which is in #3250), because unvalidated CSS can accidentally cause security or stability issues with a site.  The better approach discussed briefly would be to allow users to simply manage the values of various Bootstrap variables / colors...this would not allow any arbitrary CSS, but would allow for basic changes to the existing theme.

It was also noted, during that meeting, that these features have direct overlap with some similar concepts in DSpace-CRIS.  I'm told that DSpace-CRIS provides basic editing of Collection homepages, along with theming.  Their approach has additional validation in place & no direct CSS editing. 

Because DSpace has recently started discussing a [_potential_ merger with DSpace-CRIS](https://dspace.org/dspace-and-dspace-cris-intent-to-investigate-merger-announcement/), we are currently refraining from adding overlapping features to DSpace (at least until a merger decision has been made). 

All that said, I'm going to have to put these PRs (this one and #3250, and their corresponding backend changes) "on hold" until we can revisit them as part of the DSpace-CRIS potential merger discussions.  My apologies.

alexklbuckley commented:
> I realize I forgot to update this PR with the discussion that took place on [Jan 16 in our Developers Meeting](https://wiki.lyrasis.org/display/DSPACE/2025-01-16+DSpace+Developers+Meeting).
> 
> There was some general interest in bringing features to the User Interface that allow Administrators more control over the User Interface look & feel. However, there were concerns about allowing _editing of CSS directly_ (which is in #3250), because unvalidated CSS can accidentally cause security or stability issues with a site. The better approach discussed briefly would be to allow users to simply manage the values of various Bootstrap variables / colors...this would not allow any arbitrary CSS, but would allow for basic changes to the existing theme.
> 
> It was also noted, during that meeting, that these features have direct overlap with some similar concepts in DSpace-CRIS. I'm told that DSpace-CRIS provides basic editing of Collection homepages, along with theming. Their approach has additional validation in place & no direct CSS editing.
> 
> Because DSpace has recently started discussing a [_potential_ merger with DSpace-CRIS](https://dspace.org/dspace-and-dspace-cris-intent-to-investigate-merger-announcement/), we are currently refraining from adding overlapping features to DSpace (at least until a merger decision has been made).
> 
> All that said, I'm going to have to put these PRs (this one and #3250, and their corresponding backend changes) "on hold" until we can revisit them as part of the DSpace-CRIS potential merger discussions. My apologies.

Thanks for the update @tdonohue , appreciate that and the additional information.

Interesting that DSpace-CRIS provides basic editing of Collection homepages, along with theming.

Not a problem, thanks again for following up.
